### ISSUE TITLE
[ENH]: Support for Python 3.13

### ISSUE THREAD

**Issue by amotl** [pycaret/pycaret#4121]
### Describe the feature you want to add to this project

Hi. Coming from GH-4054, let me enumerate blockers for updating to Python 3.13 here, as discovered per CI runs on https://github.com/crate/academy-fundamentals-course/pull/23.

- https://github.com/numba/llvmlite/issues/1084
- https://github.com/alkaline-ml/pmdarima/issues/588

### Describe your proposed solution

Pick up maintenance again.

- https://github.com/pycaret/pycaret/issues/4106
- https://github.com/pycaret/pycaret/issues/4119

### Describe alternatives you've considered, if relevant

Migrate to AutoGluon, as proposed by others.

**Comment by fkiraly** [comment]
I would propose to address this by clean soft dependency management and isolation.

`pmdarima` looks like it is used only in the time series module, so we should isolate it as a soft dependency.

More generally, a review of the soft dependency strategy would be a priority I would suggest, the package has too many core dependencies and the soft dependencies are not cleanly isolated.

**Comment by amotl** [comment]
With "isolating soft dependencies", you are referring to the process of breaking out certain dependencies into "extra" dependencies, in order to make them optional for the fundamental package installation?

**Comment by fkiraly** [comment]
> With "isolating soft dependencies", you are referring to the process of breaking out certain dependencies into "extra" dependencies, in order to make them optional for the fundamental package installation?

@amotl, yes - I would go even further, having multiple, more granular sets of soft dependencies. So someone who primarily wants to use the time series module with extras does not need to install the extras for classification etc.

**Comment by amotl** [comment]
Hi. What do you think about this proposal to isolate dependencies, in order to be able permit installing and using PyCaret on most recent Python 3.13, @moezali1, @Yard1, @tvdboom, and @ngupta23?

**Comment by amotl** [comment]
This patch aims to unblock installing PyCaret on Python 3.13.

- GH-4126

**Comment by fkiraly** [comment]
In-principle sensible, but I think we should not skip the usual sequence in isolating dependencies:

1. isolate dependencies in the code, i.e., ensure the package does not break without them installed - only the feature
2. optionally, give users an advance warning that this will change
3. move the dependency to an optional dep set

Further, `pmdarima` - the only one blocking 3.13 - is imported only in two small locations, while the other time series dependencies are entagled more significantly with the code base.

I would hence suggest to start with that: https://github.com/pycaret/pycaret/issues/4127

**Comment by fkiraly** [comment]
Example on how to isolate dependencies inside the code base: https://github.com/pycaret/pycaret/pull/4129

This is bare-bones, as an example for the above procedure.

**Comment by amotl** [comment]
I've submitted a patch that permits **installation** of PyCaret on Python 3.13. However, it doesn't do anything else, neither validate if it works correctly, nor take any efforts on dependency isolation. Thanks a stack for your efforts in this regard, @fkiraly!

- https://github.com/sktime/pycaret/pull/6

**Comment by fkiraly** [comment]
I think this still updates all the versions? We should do this step-by-step.

For python 3.13, we should:

* update the bounds in `pyproject`
* relax the bounds in the `__init__` file
* add 3.13 runners to the CI, to ensure tests are executed

**Comment by jakob1379** [comment]
any progress on this?

**Comment by amotl** [comment]
Hi again. pmdarima v2.1.0 just unblocked installation on Python 3.13. Thanks a stack, @LiberiFatali and @tgsmith61591.

- https://github.com/alkaline-ml/pmdarima/issues/588#issuecomment-3528391686

**Comment by EwoutH** [comment]
I would love to be able to use PyCaret with Python 3.13!

**Comment by jakob1379** [comment]
<h3><code>pycaret</code> import across Python versions via <code>uvx</code></h3>

<h4>Command</h4>

<pre><code class="language-bash">for i in {10..14}; do
  ver="3.${i}"
  label="python${ver}"

  echo
  echo "============================================================"
  echo "TEST: import pycaret on ${label}"
  echo "CMD : uvx -p ${ver} --with pycaret python -c 'import pycaret'"
  echo "============================================================"

  if uvx -qp "${ver}" --with pycaret python -c "import pycaret"; then
    echo "RESULT: ✅ SUCCESS on ${label} (pycaret importable)"
  else
    rc=$?
    echo "RESULT: ❌ FAILURE on ${label} (exit code=${rc})"
  fi
done</code></pre>

<h4>Results</h4>

<ul>
  <li>✅ <code>python3.10</code>: import OK</li>
  <li>✅ <code>python3.11</code>: import OK</li>
  <li>❌ <code>python3.12</code>: fails at import with explicit version gate</li>
  <li>❌ <code>python3.13</code>: dependency resolution/build fails (<code>scikit-learn==1.4.2</code> → build reqs unsatisfiable)</li>
  <li>❌ <code>python3.14</code>: same as 3.13</li>
</ul>

**Comment by WilliamJudge94** [comment]
Hello @amotl, @fkiraly, @EwoutH, @jakob1379, and @kellend-git, 

I am working towards adding Python 3.13 support for Pycaret. Currently this branch (https://github.com/WilliamJudge94/pycaret/commit/77f39ae710b71969293608120c829ce15d6e21b9) passes all Python 3.13 tests (https://github.com/WilliamJudge94/pycaret/actions/runs/21196370516). 

However, it fails on 3.9 - 3.12. Before submitting a PR I will work towards maintaining compatibility with previous Python versions.

**Comment by jakob1379** [comment]
> Hello [@amotl](https://github.com/amotl), [@fkiraly](https://github.com/fkiraly), [@EwoutH](https://github.com/EwoutH), [@jakob1379](https://github.com/jakob1379), and [@kellend-git](https://github.com/kellend-git),
> 
> I am working towards adding Python 3.13 support for Pycaret. Currently this branch ([WilliamJudge94@77f39ae](https://github.com/WilliamJudge94/pycaret/commit/77f39ae710b71969293608120c829ce15d6e21b9)) passes all Python 3.13 tests (https://github.com/WilliamJudge94/pycaret/actions/runs/21196370516).
> 
> However, it fails on 3.9 - 3.12. Before submitting a PR I will work towards maintaining compatibility with previous Python versions.

what is the reason for supporting deprecated python versions? currently python3.11 is the lowers officially supported version 

https://devguide.python.org/versions/

**Comment by WilliamJudge94** [comment]
> > Hello [@amotl](https://github.com/amotl), [@fkiraly](https://github.com/fkiraly), [@EwoutH](https://github.com/EwoutH), [@jakob1379](https://github.com/jakob1379), and [@kellend-git](https://github.com/kellend-git),
> > I am working towards adding Python 3.13 support for Pycaret. Currently this branch ([WilliamJudge94@77f39ae](https://github.com/WilliamJudge94/pycaret/commit/77f39ae710b71969293608120c829ce15d6e21b9)) passes all Python 3.13 tests (https://github.com/WilliamJudge94/pycaret/actions/runs/21196370516).
> > However, it fails on 3.9 - 3.12. Before submitting a PR I will work towards maintaining compatibility with previous Python versions.
> 
> what is the reason for supporting deprecated python versions? currently python3.11 is the lowers officially supported version
> 
> https://devguide.python.org/versions/

@jakob1379 No reason other than maintaining the status quo. I was having some issues with Python 3.9, but since Python 3.10 still has security updates through 2026 I will keep that as the lowest Pycaret version.

**Comment by WilliamJudge94** [comment]
@amotl @jakob1379 the PR has been submitted with all passing tests (including the notebook tests) for Python 3.10-3.13

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
